src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageItemListViewModel.cs

```diff
@@ -640,21 +640,36 @@ private void TriggerStatusLoader()
 
         private async System.Threading.Tasks.Task ReloadPackageVersionsAsync()
         {
-            var result = await _backgroundLatestVersionLoader.Value;
-
-            await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
-
-            LatestVersion = result;
-            Status = GetPackageStatus(LatestVersion, InstalledVersion, AutoReferenced);
+            try
+            {
+                NuGetVersion latestVersion = await _backgroundLatestVersionLoader.Value;
+                await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                LatestVersion = latestVersion;
+                Status = GetPackageStatus(LatestVersion, InstalledVersion, AutoReferenced);
+            }
+            // Cancelled async operations inside of _backgroundLatestVersionLoader callpaths are expected to raise an OperationCanceledException.
+            // One shouldn't spam PostOnFailure since these are to be expected when UI interaction causes previous operations
+            // (searches, or metadata downloads or deprecationMetadata downloads, etc...) to no longer be needed.
+            catch (OperationCanceledException)
```

Comment:
Should `PostOnFailure` *completely* ignore OperationCanceledException?

Reply:
it is by design to have these async tasks cancelled. so yes, I believe it is not useful to report operationcanceledexceptions.
do you have a middle ground you are recommending for some reason?

Reply:
I think you're asking if there's a more global fix? I wondered the same. Peppering catches all over the place seems unscalable.

I can't immediately think of a scenario where you *need* that exception to bubble up.
Maybe it would require too much tracing work to go see, "ok, does this stack have a `PostOnFailure` ? Then I can't rely on the exception.

We could, by convention, *require* checking the Token directly, and not relying on the exception type.

Reply:
good question. we could do that. i don't know enough about all our uses to say. thoughts?

Reply:
@srdjanjovcic thoughts? Wondering if `PostOnFailure` excluding an exception type would be bad practice.... Say we didn't expect a Cancellation exception, and so it *should* be a failure we report, right? 

Perhaps we should first evaluate our handling of tokens, and add more catching as Rob did here.

Reply:
> Maybe it would require too much tracing work to go see, "ok, does this stack have a PostOnFailure ? Then I can't rely on the exception.

I'm not sure I follow. 
Post on failure should only be used for unawaited tasks. If you are awaiting a task, then it shouldn't have a post on failure.

Reply:
Imagine you're coding in `Method_A`', and you wonder whether you can expect `OperationCanceledException`.
You have to look inside of 'Method_B', 'Method'_C', etc, to see whether a `PostOnFailure` is used. If yes, you can't expect `OperationCanceledException`.

My point is you'd have to look at every called method and *know* about that behavior. I'm not sure it's a great practice, but it might be a convention, or we could mention it in our docs.

Reply:
> You have to look inside of 'Method_B', 'Method'_C', etc, to see whether a PostOnFailure is used. If yes, you can't expect OperationCanceledException.

This is not supposed to happen. 
If there's a post on failure, the caller never gets an exception. 
PostOnFailure should only be used when `Method_A` doesn't care about `Method_B`'s result and doesn't depend on it explicitly.

Reply:
In other words, `PostOnFailure` only applies when you write

```cs
void method(){
    JoinableTaskFactory.RunAsync(asyncTask)
    .PostOnFailure("method");
}
```

Method will not get any exceptions throw by asyncTask.

The

Reply:
this PR is about fixing these high hitters...not sure if it is good to ignore across the board. if people agree it can be, then the central fix would be best.

Reply:
> In other words, PostOnFailure only applies when you write

You will have to check every method your method calls, recursively, to find out whether `PostOnFailure` was called.

Reply:
> You will have to check every method your method calls, recursively, to find out whether PostOnFailure was called.

Why?

Reply:
i am fine to change PostOnFailure to swallow OperationCanceledExceptions centrally. If nobody has objections, (if there is no reason we need to care about these anywhere), i will do that.

Reply:
> > You will have to check every method your method calls, recursively, to find out whether PostOnFailure was called.
> 
> Why? If a method calls PostOnFailure that means you were never going to get any exception from it anyways. It's a fire and forget method.

Let's have a chat offline. Seems we are miscommunicating on this

Reply:
> i am fine to change PostOnFailure to swallow OperationCanceledExceptions centrally. If nobody has objections, (if there is no reason we need to care about these anywhere), i will do that.

I was hoping Srdjan would be able to chime-in before making such broad impacting decisions.

Reply:
> i am fine to change PostOnFailure to swallow OperationCanceledExceptions centrally. If nobody has objections, (if there is no reason we need to care about these anywhere), i will do that.

I have an objection to this.

HttpClient's timeouts are exposed via `TaskCancelledException`. If NuGet.Protocol is doesn't retry, or exceeds the maximum number of retries, and therefore `TaskCencelledException` is bubbled up to PM UI, we want to ensure we handle it correctly, and not just tell the customer "waiting for response" forever. If we don't handle this correctly and we get an unhandled `TaskCancelledException` in a file and forget method in VS, we shouldn't systematically ignore it. We need the fault telemetry to tell us customers are getting a poor experience.

I know it's much more effort to handle task cancellation, and exception handling, "correctly". But I think that's what we should be striving for. For VS components that use `CancellationTokenSource`, it should catch `TaskCancelledException` and PostFault when the `CancellationTokenSource` is not in the cancelled state, because it means an unexpected cancellation happened and silently ignoring it might have negative customer impact.

Reply:
@zivkan  - does it make sense to centralize that handler in PostAsFailure()? Seems reasonable to not catch too widely, and perhaps checking the token source for actually being cancelled is the right thing.

Then, we would only need to go make sure our codepaths, perhaps in order of failures posted, ensure that they plumb the token down correctly.

Reply:
I'm not so concerned about how it's implemented. I just want to avoid blindly filtering out all TaskCancelledException, especially when it leads to a bad customer experience (VS reports it's waiting for a response that will never arrive).

Reply:
> Then, we would only need to go make sure our codepaths, perhaps in order of failures posted, ensure that they plumb the token down correctly.

Isn't there a second step too, where you'd have to find whether `PostOnFailure` is used (I was trying to voice this concern in the beginning of this thread).

> I just want to avoid blindly filtering out all TaskCancelledException, especially when it leads to a bad customer experience (VS reports it's waiting for a response that will never arrive).

Yes and to know this, I have to know if *any* codepath invokes `PostOnFailure`, and if it may possibly swallow an exception I need way up in the UI. 

I would really prefer some sort of convention or pattern be used. Maybe a fluent syntax, or even naming conventions for methods/classes that somehow indicates Cancellation does not bubble up?

Reply:
> I'm not so concerned about how it's implemented. I just want to avoid blindly filtering out all TaskCancelledException, especially when it leads to a bad customer experience (VS reports it's waiting for a response that will never arrive).

Can you link to the part that where an actual problem is articulated?

My understanding is that this behavior is completely benign.

Reply:
Here's the first example I found: https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs#L895

Reply:
Is that a response my question?
I'm not sure how that answers it. 
There's no fire and forget here, the example explicitly checks for a cancellation exception.

Reply:
Yes it's a response to the question of where does VS expect these cancellations so it can handle them, and possibly have a bad customer experience (from the quote you gave from zivkan).

> There's no fire and forget here, the example explicitly checks for a cancellation exception.

True, but my point is you'd have to know how all called methods handle PostOnFailure.

I suppose only if there's a spot where we pass a `CancellationToken` down, and then that token is used in a `PostOnFailure`, could we unknowingly lose expected cancellation exceptions. Is that your thinking?

Reply:
> could we unknowingly lose expected cancellation exceptions

The exception is lost anyways, this is only about whether we get a telemetry event for it.

Reply:
> Should PostOnFailure completely ignore OperationCanceledException

I think we all have different interpretations of this. Maybe a quick offline chat with the 3+ of us can resolve?

Reply:
i am going to spend a bit more time in this issue than i had planned due to other changed plans. 

here is telemetry of 16.10 usage where exception contains "Canceled':
![image](https://user-images.githubusercontent.com/8865080/108119251-5d4fb280-7054-11eb-851f-06bb181aaf7d.png)

going to go dig a bit into this.
let me go look at what it would take to fix all these code paths. etc...
we may have a chat, but lets pause this debate for now.